### PR TITLE
feat(headless): expose relay on the engine

### PR DESCRIPTION
**Pull request by louis-bompart**
- Refactor the configureAnalytics & getAnalyticsSource
  - Make them into memoized selectors to shave off some computing
- Exposes Relay onto the engine.

KIT-3018

### CODE REVIEW

File: packages/headless/src/api/analytics/analytics-relay-client.test.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  it('creates a Relay client properly and returns it', () => {
+    const state = createMockState();
+
+    const relay = getRelayInstanceFromState(state);
+
+    expect(mockedCreateRelay).toHaveBeenCalledWith({
+      url: state.configuration.analytics.nextApiBaseUrl,
+      token: state.configuration.accessToken,
+      trackingId: state.configuration.analytics.trackingId,
+      source: expect.arrayContaining([]),
```
Comment (olamothe) [inline packages/headless/src/api/analytics/analytics-relay-client.test.ts:31]: Shouldn't there be something in the array ? Like `@coveo/headless@unit-testing-version` (from createMockState()) ?

File: packages/headless/src/app/engine.test.ts [hunk 2]
```diff
@@ -259,4 +261,23 @@ describe('engine', () => {
...
   });
+
+  it('should exposes a Relay instance', () => {
+    const engine = initEngine();
+    expect(engine.relay).toBeDefined();
+  });
+
+  it('should return the same instance of Relay if the state has not changed', () => {
+    const engine = initEngine();
+    expect(engine.relay).toBe(engine.relay);
```
Comment (olamothe) [inline packages/headless/src/app/engine.test.ts:272]: That seems like a useless test 🤔  ?

Comment (louis-bompart) [reply packages/headless/src/app/engine.test.ts:272]: Not really: This test shows that if we call the relay getter twice for a given state, we get the same instance of Relay.

File: packages/headless/src/api/analytics/analytics-relay-client.test.ts (hunk 1 above)
Comment (louis-bompart) [reply packages/headless/src/api/analytics/analytics-relay-client.test.ts]: 👍 tested through isolation

File: packages/headless/src/app/engine.test.ts (hunk 2 above)
Comment (olamothe) [reply packages/headless/src/app/engine.test.ts:272]: Ok I understand. 
Then maybe just a small comment that explains this would help future readers

**Review by olamothe — APPROVED** [review]

File: packages/headless/src/app/engine.test.ts [hunk 3]
```diff
@@ -259,4 +261,23 @@ describe('engine', () => {
       });
     });
   });
+
+  it('should exposes a Relay instance', () => {
+    const engine = initEngine();
+    expect(engine.relay).toBeDefined();
+  });
+
+  it('should return the same instance of Relay if the state has not changed', () => {
```
Comment (louis-bompart) [inline packages/headless/src/app/engine.test.ts:270]:
```suggestion
  it('should return the exact same instance of Relay if the state has not changed between two calls to engine.relay', () => {
```

**Review by Spuffynism — APPROVED** [review]

**Pull request merged**